sdk/storage/azure-storage-blob-cryptography/src/test/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlockBlobAsyncApiTests.java

```diff
@@ -0,0 +1,1920 @@
+                new ParallelTransferOptions().setProgressListener(mockListener),
+                new DownloadRetryOptions().setMaxRetryRequests(3), null, false));
+
+        StepVerifier.create(response)
+            .expectNextCount(1)
+            .verifyComplete();
+
+        // We should receive at least one notification reporting an intermediary value per block, but possibly more
+        // notifications will be received depending on the implementation. We specify numBlocks - 1 because the last
+        // block will be the total size as above. Finally, we assert that the number reported monotonically increases.
+        if (numBlocks == 0) {
+            numBlocks++;
+        }
+
+        assertTrue(progress.stream().filter(it -> it != file.length()).count() >= numBlocks - 1);
+
+        // Should receive at least one notification indicating completed progress, multiple notifications may be
+        // received if there are empty buffers in the stream.
+        assertTrue(progress.stream().anyMatch(it -> it == fileSize));
+
+        // There should be NO notification with a larger than expected size.
+        assertTrue(progress.stream().noneMatch(it -> it > fileSize));
+
+        Files.deleteIfExists(file.toPath());
+        Files.deleteIfExists(outFile.toPath());
+    }
+
+    @Test
+    public void downloadRequiresEncryption() {
+        String blobName = bec.getBlobName();
+        getBlobClientBuilder(ENV.getPrimaryAccount().getCredential(), cc.getBlobContainerUrl())
+            .blobName(blobName)
+            .buildClient()
+            .upload(DATA.getDefaultBinaryData());
+
+        bec = getEncryptedClientBuilder(fakeKey, null, ENV.getPrimaryAccount().getCredential(),
+            cc.getBlobContainerUrl())
+            .blobName(blobName)
+            .requiresEncryption(true)
+            .buildEncryptedBlobAsyncClient();
+
+        StepVerifier.create(bec.download())
+            .verifyError(IllegalStateException.class);
+
+        StepVerifier.create(bec.downloadWithResponse(null, null, null, false))
+            .verifyError(IllegalStateException.class);
+    }
+
+    @Test
+    public void encryptionUploadISOverwriteFails() {
+        //The error thrown diverges from the sync client, which throws a BlobStorageException, but chaging it would
+        //be a breaking change.
+        StepVerifier.create(bec.upload(DATA.getDefaultBinaryData()).then(bec.upload(DATA.getDefaultBinaryData())))
+            .verifyError(IllegalArgumentException.class);
+    }
+
+    @Test
+    public void encryptionUploadISOverwrite() {
+        Mono<byte[]> response = bec.upload(BinaryData.fromBytes(getRandomByteArray(Constants.KB)))
+            .then(bec.upload(DATA.getDefaultFlux(), null, true)
+                .then(FluxUtil.collectBytesInByteBufferStream(bec.downloadStream())));
+
+        StepVerifier.create(response)
+            .assertNext(r -> assertArrayEquals(DATA.getDefaultBytes(), r))
+            .verifyComplete();
+    }
+
```

Comment:
Comparing with the sync tests, there is an additional test `encryptionUploadISSyncNotANoop`. Do we need to port this over to an async version? Seems like we'd want it assuming we don't already have coverage that ensures our async upload is actually encrypting blobs.